docs/collections.md

```diff
@@ -0,0 +1,271 @@
+
+Use `addAsCompletion($collection)` in place of `addAsRollback($collection)`, or implement \Robo\Contract\CompletionInterface. Completions otherwise work exactly like rollbacks.
+
+## Temporary Objects
+
+Since the concept of temporary objects that are cleaned up  on failure is a common pattern, Robo provides built-in support for them. Temporary directories and files are provided out of the box; other kinds of temporary objects can be easily created using the Temporary global collection.
+
+### Temporary Directories
+
+It is recommended that operations that perform multiple filesystem operations should, whenever possible, do most of their work in a temporary directory that is moved into its final location near the end of the task collection. This makes it easy to clean up after an aborted task.
+
+If you use a temporary directory created by `$this->taskTmpDir()`, it will automatically be removed when the collection completes or rolls back. Move the temporary directory to another location to prevent its deletion.
+
+``` php
+<?php
+class RoboFile extends \Robo\Tasks
+{
+    function myOperation()
+    {
+        $collection = $this->collection();
+        
+        // Create a temporary directory, and fetch its path.
+        $work = $this->taskTmpDir()
+          ->addToCollection($collection)
+          ->getPath();
+
+        $this->taskOther($work)
+          ->addToCollection($collection);
+
+        // If all of the tasks succeed, then rename the temporary directory
+        // to its final name.
+        $this->taskFileSystemStack()
+          ->rename($work, 'destination')
+          ->addToCollection($collection);
+        
+        $result = $collection->run();
+    }
+}
+?>
+```
+
+In the previous example, the path to the temporary directory is stored in the variable `$work`, and is passed as needed to the parameters of the other tasks as they are added to the collection. After the task collection is run, the temporary directory will be automatically deleted. In the example above, the temporary directory is renamed by the last task in the collection. This allows the working directory to persist; the collection will still attempt to remove the working directory, but no errors will be thrown if it no longer exists in its original location. Following this pattern allows Robo scripts to easily and safely do work that cleans up after itself on failure, without introducing a lot of branching or additional error recovery code.
+
+### Temporary Files
+
+Robo also provides an API for creating temporary files. They may be created via `$this->taskTmpFile()`; they are used exactly like `$this->taskWrite()`, except they are given a random name on creation, and are deleted when their collection completes.  If they are not added to a collection, then they are deleted when the script terminates.
+
+### The Temporary Global Collection
```

Comment:
I tried to read this chapter 3 times and for me it it is hard to understand the problem being solved.
could you provide an example here?